numeric_ranges_algorithms/numeric_ranges_algorithms.md

```diff
@@ -0,0 +1,427 @@
+
+## Input and output ranges
+
+We follow the approach of P3179R8.  For existing non-ranges algorithms that take iterator pairs as input and return a value, their ranges versions just return a value.  This covers `reduce`.  The non-ranges versions of `inclusive_scan` and `exclusive_scan` take an input range and an output iterator, and return an iterator to the element past the last element written.  This works more or less like `ranges::transform`, so we imitate the approach in P3179R8 by having those algorithms take an input range and an output range (both sized), and return an alias to `in_out_result`.  (If the output doesn't suffice to contain the input, then callers need to know what part of the input hasn't yet been processed.)  These algorithms have freedom to reorder binary operations and can copy the binary operator, so there is no value in returning it (as a way for callers to get at any of its modified state).
+
+## Constexpr parallel algorithms?
+
+<a href="https://wg21.link/p2902r1">P2902R1</a> proposes to add `constexpr` to the parallel algorithms.  P3179R8 does not object to this; see Section 2.10.  We continue the approach of P3179R8 in not opposing P2902R1's approach, but also not depending on it.
+
+# Implementation
+
+TODO
+
+The oneDPL library has deployment experience.
+
+# Wording
+
+> Text in blockquotes is not proposed wording, but rather instructions for generating proposed wording.
+>
+> Assume that <a href="https://isocpp.org/files/papers/P3179R8.html">P3179R8</a> has been applied to the Working Draft.
+
+## Update feature test macro
+
+> In [version.syn], increase the value of the `__cpp_lib_parallel_algorithm` macro by replacing YYYMML below with the integer literal encoding the appropriate year (YYYY) and month (MM).
+
+```c++
+#define __cpp_lib_parallel_algorithm YYYYMML // also in <algorithm>
+```
+
+## Change [numeric.ops.overview]
+
+> Change [numeric.ops.overview] (the `<numeric>` header synopsis) as follows.
+
+### Add declaration of exposition-only concepts
+
+> Add declarations of exposition-only concepts _`indirectly-binary-foldable-impl`_ and _`indirectly-binary-foldable`_ to [numeric.ops.overview] (the `<numeric>` header synopsis) as follows.
+
+```
+    template<class F, class T, class I, class U>
+      concept @_indirectly-binary-left-foldable-impl_@ =  // @_exposition only_@

```

Comment:
```suggestion
      concept @_indirectly_binary_left_foldable-impl_@ =  // @_exposition only_@

```

Reply:
nit: Underscores over "-"  looks better.

Reply:
@abhilash1910 wrote:
> nit: Underscores over "-" looks better.

The Standard convention for exposition-only concepts is to use kabob-case (hyphens, not underscores.